### ISSUE TITLE
Using transactions to store arbitrary datums is not allowed

### ISSUE THREAD

**Issue by evgeny-osipenko** [IntersectMBO/cardano-ledger#2688]
Currently, we are working on a Plutus project, that is intended to work like this:
- Some amount of tokens is locked behind a script on Cardano.
- Participants communicate off-chain to coordinate the usage of these tokens.
- A proof of coordination is a document, signed by all participants, with their corresponding pubkeys stored in the script.
- When given such a proof, the script should allow the action, described in the document, to proceed (withdraw a certain amount of tokens or close the on-chain contract entirely and split the tokens between participants).
- The contract is persistent: the document may allow to only withdraw a fraction of the locked tokens, while the rest should be kept protected by the same script, albeit with a different internal state.
- As a failsafe, if some participants refuse to cooperate, a different kind of document (a balance proof) can be used to initiate a termination of the contract. Unless another participant submits a newer version of this document in the allotted time, the script will close the contract and split the tokens according to the balance proof.

For this to work, our script needs to verify the signatures of the submitted documents.

Initially, we decided to add these documents as supplementary datums in the transaction, and provide the signed hash in the redeemer: the script would use an intrinsic to verify the signature, and then perform a lookup in the transaction witnesses. However, when deploying our application into the testnet, we discovered that such usage is unsupported: unless a datum is directly used in an output, the transaction is rejected.

Is there a good reason for this check that I am not aware of? Is it possible to relax this rule and allow any datums to be included in a transaction, specifically for use-case scenarios such as ours? Or is there a better way to solve the problem described above (verifying signatures of arbitrary documents)?

**Comment by JaredCorduan** [comment]
A lot of times we use hashes as pointers, for practical purposes (making things smaller, separating witnesses from the body, etc). In an ideal world where we did not care about space, we would inline these things, such as the data inside of transaction outputs. We would like to keep open the option of converting between the practical structures and the idealized ones. This is why we chose to not allow data hashes for things that do not correspond to something inside of the given transaction.

I believe Plutus will be adding new builtins for the use case you mention, I know many developers would like such a thing.
